logic/scorer/scorer.go

```diff
@@ -87,29 +103,127 @@ func scoreFifteens(ptVals []int) (scoreType, int) {
 		return 0
 	}
 
 	var many uint
-	for i, o := range ptVals {
+	for i := start; i < cardsPerHand; i++ {
+		o := ptVals[i]
 		if o > goal {
-			break
+			continue
 		} else if o == goal {
 			many++
 		} else {
 			// o is less than the goal. See what we can find with it
-			subWith := howManyAddUpTo(goal-o, ptVals[i+1:])
+			subWith := howManyAddUpTo(goal-o, ptVals, i+1)
 			many += subWith
 		}
 	}
 
 	return many
 }
 
-// Assumes input is sorted and has len 5
-func scoreRunsAndPairs(values []int) (scoreType, int) { //nolint:gocyclo
+func scorePairs(valuesToCounts valueToCount) (scoreType, int) {
+	pairPoints := 0
+	var pairType scoreType
+	for n, ct := range valuesToCounts {
+		if n == 0 {
+			continue
+		}
+		switch ct {
+		case 4:
+			return quad, 12
+		case 3:
+			pairType |= triplet
+			pairPoints += 6
+		case 2:
+			pairPoints += 2
+			if pairType&onepair > 0 {
+				return twopair, 4
+			}
+			pairType |= onepair
+		}
+	}
+	if pairType == 0 {
+		return none, 0
+	}
+	return pairType, pairPoints
+}
+
+func scoreRuns(values values, valuesToCounts valueToCount) (scoreType, int) {
+	var (
+		mult   uint8
+		runLen uint8
+	)
+	for _, v := range values {
+		if _, ok := valuesToCounts.get(v - 1); ok {
+			// this is already part of a run; skip calculation
+			continue
+		}
+		runLen = 1
+		mult, _ = valuesToCounts.get(v)
+		// we're at the potential beginning of a run
+		for nextUp := v + 1; valuesToCounts.containsKey(nextUp); nextUp++ {
+			ct, _ := valuesToCounts.get(nextUp)
+			mult *= ct
+			runLen++
+		}
+		if runLen >= 3 {
+			return calculateTypeAndPoints(runLen, mult)
+		}
+	}
+	return none, 0
+}
+
+func calculateTypeAndPoints(longest, mult uint8) (scoreType, int) {
+	if longest < 3 {
+		return none, 0
+	}
+	if longest == 5 {
+		return run5, 5
+	}
+	// typeMap maps the run length and multiplier to a scoring type
+	typeMap := [5][5]scoreType{
+		3: {1: run3, 2: doubleRunOfThree, 3: tripleRunOfThree, 4: doubleDoubleRunOfThree},
+		4: {1: run4, 2: doubleRunOfFour},
+	}
+	return typeMap[longest][mult], int(longest) * int(mult)
+}
+
+func resolveScoreType(st scoreType) scoreType {
+	// this mask is generated as following:
+	// 00011111
+	//    ^ doubleRunOfThree bit
+	// 00000011
+	//      ^ tripleRunOfThree bit (minus one to get the two bits below)
+	// 11111100 & 00011111 = 00011100, with the three bits set being the runs that include pairs
+	runsWithPairsMask := ^(tripleRunOfThree - 1) & ((doubleRunOfThree << 1) - 1)
+	if runsWithPairs := runsWithPairsMask & st; runsWithPairs > 0 {
+		// i.e. a triple run of three is scored without the triplet
+		return runsWithPairs
+	}
+	if st > none {
+		return st & ^1 // clear off the none bit if it's set
+	}
+	return none
+}
+
+func scoreRunsAndPairsV2(values values) (scoreType, int) {
+	var valuesToCounts valueToCount
+	for _, v := range values {
+		valuesToCounts[v]++
+	}
+	pairType, pairPts := scorePairs(valuesToCounts)
+	runType, runPts := scoreRuns(values, valuesToCounts)
```

Comment:
:microscope: currently, this will iterate through `values` twice (once straight through, and the second time will also iterate a run, if it exists) and `valuesToCounts` once.

:mage: Can we score runs and pairs with only one iteration through `values` and one iteration through `valuesToCounts`?